tools/check

```diff
@@ -218,6 +219,84 @@ run_test() {
     flutter test
 }
 
+# Check the Flutter version in pubspec.yaml is commented with a commit ID,
+# which agrees with the version, and the commit is from upstream main.
+run_flutter_version() {
+    # Omitted from this files check:
+    #   tools/check
+    files_check pubspec.yaml \
+    || return 0
+
+    local flutter_tree flutter_git
+    flutter_tree=$(flutter_tree)
+    flutter_git=( git --git-dir="${flutter_tree}"/.git )
+
+    # Parse our Flutter version spec and its commit-ID comment.
+    local parsed flutter_version flutter_commit
+    # shellcheck disable=SC2207 # output has controlled whitespace
+    parsed=( $(
+        perl <pubspec.yaml -0ne '
+             print "$1 $2" if (
+                 /^  sdk: .*\n  flutter: '\''>=(\S+)'\''\s*# ([0-9a-f]{40})$/m)'
+    ) ) || return
+    if [ -z "${#parsed[@]}" ]; then
+        echo >&2 "error: Flutter version spec not recognized in pubspec.yaml"
+        return 1
+    fi
+    flutter_version="${parsed[0]}"
+    flutter_commit="${parsed[1]}"
+
+    # Check the version name matches the commit ID.
+    local commit_described predicted_version
+    commit_described=$(
+        "${flutter_git[@]}" describe --tags "${flutter_commit}"
+    ) || return
+    predicted_version=$(
+        echo "${commit_described}" \
+        | perl -lne 'print if (s
+             # This transformation is ad hoc.
+             # If we find cases where it fails, we can study
+             # how the `flutter` tool actually decides the version name.
+             <^(\d+\.\d+\.\d+-) (\d+) (\.\d+\.pre) -(\d+) -g[0-9a-f]+$>
+             <$1 . ($2 + 1) . $3 . "." . $4>xe)'
+    ) || return
+    if [ -z "${predicted_version}" ]; then
+        cat >&2 <<EOF
+error: unexpected 'git describe' result on Flutter commit in pubspec.yaml
+Commit ${flutter_commit} was described as: ${commit_described}
+EOF
+        return 1
+    fi
+    if [ "${flutter_version}" != "${predicted_version}" ]; then
+        cat >&2 <<EOF
+error: Flutter commit in pubspec.yaml seems to differ from version bound
+Commit ${flutter_commit} was described as: ${commit_described}
+We therefore expect the Flutter version name to be: ${predicted_version}
+But the Flutter version bound in pubspec.yaml is: ${flutter_version}
+EOF
+        return 1
+    fi
+
+    # Check the commit is an acceptable commit.
+    local commit_count
+    commit_count=$(
+        "${flutter_git[@]}" rev-list --count origin/main.."${flutter_commit}"
```

Comment:
I've named the remote "upstream" instead of "origin", so "origin/main" doesn't work for me (here and below). I've named it back to "origin" for testing this new code; should I keep it that way or should the code adapt to handle the "upstream" name?

Reply:
I'd suggest just switching back to "origin". It'd be possible to make this handle both conventions, but would make the logic more complicated.

I don't think the Flutter install steps do anything to override Git's default here, so most contributors will have "origin".

Reply:
(Personally I switched back from using "upstream" to using "origin" a little while ago, for Git repos in general, partly just because I find "origin" is quicker for me to reliably type — "upstream" has long runs on the same hand.)